### PR TITLE
Sync ``libasr`` from LPython

### PR DESCRIPTION
**Pull request by czgdp1807**
Corresponding LPython PR - https://github.com/lcompilers/lpython/pull/1670

### CODE REVIEW

File: src/libasr/alloc.h
Comment (czgdp1807) [inline src/libasr/alloc.h:1]: Note - Changes done as per https://github.com/lcompilers/lpython/pull/1670#discussion_r1159346226

File: src/libasr/alloc.h [hunk 1]
```diff
@@ -58,6 +58,7 @@ class Allocator
             LCOMPILERS_ASSERT(start != nullptr);
             size_t addr = current_pos;
             current_pos += align(s);
+            if (size_current() > size_total()) throw std::bad_alloc();
```
Comment (certik) [inline src/libasr/alloc.h:61]:
```suggestion
```

**Review by Thirumalai-Shaktivel — APPROVED** [review]
LGTM, Thank you!

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -8035,7 +8070,7 @@ Result<std::unique_ptr<LLVMModule>> asr_to_llvm(ASR::TranslationUnit_t &asr,
     pass_manager.apply_passes(al, &asr, pass_options, diagnostics);
 
     // Uncomment for debugging the ASR after the transformation
-    // std::cout << LFortran::pickle(asr, true, true, true) << std::endl;
+    // std::cout << LCompilers::LPython::pickle(asr, false, false, false) << std::endl;
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_llvm.cpp:8073]:
```suggestion
    // std::cout << LCompilers::LFortran::pickle(asr, false, false, false) << std::endl;
```

or 

```suggestion
    // std::cout << LFortran::pickle(asr, false, false, false) << std::endl;
```

**Review by Pranavchiku — APPROVED** [review]
Looks good to me! Thank you very much for doing this.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 2 above)
Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:8073]: I wouldn't worry too much about this. Instead I would like to have an option, `--pass=all` which applies all the passes and then show the ASR. Then we won't be needing the above `std::cout` line at all.

Comment (Pranavchiku) [reply src/libasr/codegen/asr_to_llvm.cpp:8073]: Why apply for all passes again when `asr` contains the required content?

Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_llvm.cpp:8073]: Yup, I'm +1 on it. It will be very helpful for debugging!

Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:8073]: > Why apply all passes again when `asr` contains the required content?

Debugging purposes. `--pass=all --show-asr` will show you how the ASR looks like after applying all the passes.

Comment (Pranavchiku) [reply src/libasr/codegen/asr_to_llvm.cpp:8073]: I mean to say `asr` is the ASR generated after applying all passes listed in `pass_manager.h`.

Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_llvm.cpp]: Nope. I think, the option: `--show-asr` only prints the ASR generated by the `ast_to_asr.cpp`. The ASR passes are applied in the Backend: LLVM
https://github.com/lfortran/lfortran/blob/3fc8c57f605e47cfc6f01aacbdcf85a3d6317709/src/libasr/codegen/asr_to_llvm.cpp#L8027-L8035

**Review by Smit-create — APPROVED** [review]

**Review by Shaikh-Ubaid — APPROVED** [review]

File: src/libasr/codegen/asr_to_c.cpp [hunk 3]
```diff
@@ -1208,42 +1267,27 @@ R"(
...
     }
 
-    void visit_GoTo(const ASR::GoTo_t &x) {
-        std::string indent(indentation_level*indentation_spaces, ' ');
-        src =  indent + "goto " + std::string(x.m_name) + ";\n";
-    }
-
-    void visit_GoToTarget(const ASR::GoToTarget_t &x) {
-        src = std::string(x.m_name) + ":\n";
-    }
```
Comment (certik) [inline src/libasr/codegen/asr_to_c.cpp:1231]: These were introduced in 6345cb21229eaeb451aeee8aa9c77ba1327f0364.

Comment (certik) [reply src/libasr/codegen/asr_to_c.cpp:1231]: I see, they were refactored in https://github.com/lcompilers/lpython/pull/1636.

**Review by certik — APPROVED** [review]
I think this is fine.

**Pull request merged**